### PR TITLE
Use configurations to choose the TLS provider.

### PR DESCRIPTION
**Pull request by s-ludwig**
This sidesteps dub/issues#1089 and makes vibe.d usable without linking against OpenSSL again.

This is an alternative approach to #1769.

### CODE REVIEW

File: stream/dub.sdl [hunk 1]
```diff
@@ -14,19 +12,25 @@ importPaths "."
 copyFiles "../lib/win-i386/libeay32.dll" "../lib/win-i386/ssleay32.dll" platform="windows-x86"
 copyFiles "../lib/win-amd64/libeay32.dll" "../lib/win-amd64/ssleay32.dll" platform="windows-x86_64"
 
-configuration "windows-mscoff" {
+// LDC 1.0.0 doesn't include libz in libphobos
+libs "z" platform="posix-ldc"
+libs "z" platform="gdc"
+
+configuration "openssl-mscoff" {
```
Comment (wilzbach) [inline stream/dub.sdl]: Hehe, I would this doesn't result in combinatoric explosion in the future, but I guess the "hack" would be to move SSL into separate sub-package if more configurations options are ever needed for `stream`. Maybe - on the long run - it makes sense to go an approach like [SecureD](https://github.com/LightBender/SecureD) and provide a common API over botan/openssl/libressl/... in a new "ssl" package?

Comment (s-ludwig) [inline stream/dub.sdl]: Yeah, the next step would definitely be to separate this into a vibe-d:tls package. Maybe I should even better do that now... Having an abstraction independent of vibe.d's streams also makes sense in principle. It should of course also be noted that there are two other obvious TLS implementation choices, namely the ones provided by the OS on Windows and macOS.

**Comment by s-ludwig** [comment]
Great, this is messed up. AppVeyor doesn't react anymore. The "push" build on Travis seems to be merged into master instead of 0.8.0, which makes the version test fail. And finally the remaining Travis build fails in a way that I can't reproduce locally. :-1:

**Comment by wilzbach** [comment]
> Great, this is messed up. AppVeyor doesn't react anymore. The "push" build on Travis seems to be merged into master instead of 0.8.0, which makes the version test fail. And finally the remaining Travis build fails in a way that I can't reproduce locally. :-1:

Btw (in case you haven't enabled this), there's at least the option to auto-cancel "old" commit / builds on Travis:

https://blog.travis-ci.com/2017-03-22-introducing-auto-cancellation

I think on AppVeyor one needs to cancel the builds manually.

> And finally the remaining Travis build fails in a way that I can't reproduce locally.

I can:

```
> dub -v build --combined -b release --compiler=dmd --config=libevent
Using dub registry url 'http://code.dlang.org/'
Refreshing local packages (refresh existing: true)...
Looking for local package map at /var/lib/dub/packages/local-packages.json
Looking for local package map at /home/seb/.dub/packages/local-packages.json
Try to load local package map at /home/seb/.dub/packages/local-packages.json
Determined package version using GIT: eventcore 0.8.8+commit.1.g420c361
Determined package version using GIT: hb-web 0.0.1+commit.3.gf97d88f
Determined package version using GIT: hb-ddb 0.2.5+commit.38.gb515f6b
Determined package version using GIT: vibe-core 1.0.0-beta.6+commit.4.g698824e
Determined package version using GIT: iopipe ~master
Determined package version using GIT: oauth 0.2.0-beta.2+commit.1.gf054671
Determined package version using GIT: inifiled 1.0.2
Determined package version using GIT: dscanner 0.4.0+commit.67.ga853936
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Ignoring version specification (>=0.0.0) for path based dependency .
Determined package version using GIT: vibe-d 0.8.0-rc.4+commit.2.g6ae5c111
No package found in current working directory.
Refreshing local packages (refresh existing: false)...
Looking for local package map at /var/lib/dub/packages/local-packages.json
Looking for local package map at /home/seb/.dub/packages/local-packages.json
Try to load local package map at /home/seb/.dub/packages/local-packages.json
  Error getting sub package vibe-d:redis: Sub package "vibe-d:redis" doesn't exist.
  Found dependency vibe-d:data 0.8.0-rc.4+commit.2.g6ae5c111
    Found dependency vibe-d:utils 0.8.0-rc.4+commit.2.g6ae5c111
  Found dependency vibe-d:core 0.8.0-rc.4+commit.2.g6ae5c111
    Found dependency libevent 2.0.2+2.0.16
    Found dependency libasync 0.8.3
      Found dependency memutils 0.4.9
    Found dependency vibe-core 1.0.0
      Found dependency eventcore 0.8.12
        Found dependency taggedalgebraic 0.10.7
  Error getting sub package vibe-d:textfilter: Sub package "vibe-d:textfilter" doesn't exist.
  Error getting sub package vibe-d:crypto: Sub package "vibe-d:crypto" doesn't exist.
  Error getting sub package vibe-d:mongodb: Sub package "vibe-d:mongodb" doesn't exist.
  Error getting sub package vibe-d:inet: Sub package "vibe-d:inet" doesn't exist.
  Error getting sub package vibe-d:diet: Sub package "vibe-d:diet" doesn't exist.
  Error getting sub package vibe-d:web: Sub package "vibe-d:web" doesn't exist.
  Error getting sub package vibe-d:tls: Sub package "vibe-d:tls" doesn't exist.
  Found dependency vibe-d:stream 0.8.0-rc.4+commit.2.g6ae5c111
    Found dependency openssl 1.1.5+1.0.1g
    Found dependency botan 1.12.9
      Found dependency botan-math 1.0.3
  Error getting sub package vibe-d:http: Sub package "vibe-d:http" doesn't exist.
  Error getting sub package vibe-d:mail: Sub package "vibe-d:mail" doesn't exist.
Refreshing local packages (refresh existing: false)...
Looking for local package map at /var/lib/dub/packages/local-packages.json
Looking for local package map at /home/seb/.dub/packages/local-packages.json
Try to load local package map at /home/seb/.dub/packages/local-packages.json
  Error getting sub package vibe-d:redis: Sub package "vibe-d:redis" doesn't exist.
  Found dependency vibe-d:data 0.8.0-rc.4+commit.2.g6ae5c111
    Found dependency vibe-d:utils 0.8.0-rc.4+commit.2.g6ae5c111
  Found dependency vibe-d:core 0.8.0-rc.4+commit.2.g6ae5c111
    Found dependency libevent 2.0.2+2.0.16
    Found dependency libasync 0.8.3
      Found dependency memutils 0.4.9
    Found dependency vibe-core 1.0.0
      Found dependency eventcore 0.8.12
        Found dependency taggedalgebraic 0.10.7
  Error getting sub package vibe-d:textfilter: Sub package "vibe-d:textfilter" doesn't exist.
  Error getting sub package vibe-d:crypto: Sub package "vibe-d:crypto" doesn't exist.
  Error getting sub package vibe-d:mongodb: Sub package "vibe-d:mongodb" doesn't exist.
  Error getting sub package vibe-d:inet: Sub package "vibe-d:inet" doesn't exist.
  Error getting sub package vibe-d:diet: Sub package "vibe-d:diet" doesn't exist.
  Error getting sub package vibe-d:web: Sub package "vibe-d:web" doesn't exist.
  Error getting sub package vibe-d:tls: Sub package "vibe-d:tls" doesn't exist.
  Found dependency vibe-d:stream 0.8.0-rc.4+commit.2.g6ae5c111
    Found dependency openssl 1.1.5+1.0.1g
    Found dependency botan 1.12.9
      Found dependency botan-math 1.0.3
  Error getting sub package vibe-d:http: Sub package "vibe-d:http" doesn't exist.
  Error getting sub package vibe-d:mail: Sub package "vibe-d:mail" doesn't exist.
Checking for missing dependencies.
External sub package vibe-d:redis 0.8.0-rc.4+commit.2.g6ae5c111 not found.
Using fixed selection libevent 2.0.2+2.0.16
Using fixed selection libasync 0.8.3
Using fixed selection memutils 0.4.9
Using fixed selection vibe-core 1.0.0
Using fixed selection eventcore 0.8.12
Using fixed selection taggedalgebraic 0.10.7
External sub package vibe-d:textfilter 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:crypto 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:mongodb 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:inet 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:diet 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:web 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:tls 0.8.0-rc.4+commit.2.g6ae5c111 not found.
Using fixed selection openssl 1.1.5+1.0.1g
Using fixed selection botan 1.12.9
Using fixed selection botan-math 1.0.3
External sub package vibe-d:http 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:mail 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:redis 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:textfilter 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:crypto 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:mongodb 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:inet 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:diet 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:web 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:tls 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:http 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:mail 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:redis 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:textfilter 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:crypto 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:mongodb 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:inet 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:diet 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:web 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:tls 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:http 0.8.0-rc.4+commit.2.g6ae5c111 not found.
External sub package vibe-d:mail 0.8.0-rc.4+commit.2.g6ae5c111 not found.
Refreshing local packages (refresh existing: false)...
Looking for local package map at /var/lib/dub/packages/local-packages.json
Looking for local package map at /home/seb/.dub/packages/local-packages.json
Try to load local package map at /home/seb/.dub/packages/local-packages.json
  Error getting sub package vibe-d:redis: Sub package "vibe-d:redis" doesn't exist.
  Found dependency vibe-d:data 0.8.0-rc.4+commit.2.g6ae5c111
    Found dependency vibe-d:utils 0.8.0-rc.4+commit.2.g6ae5c111
  Found dependency vibe-d:core 0.8.0-rc.4+commit.2.g6ae5c111
    Found dependency libevent 2.0.2+2.0.16
    Found dependency libasync 0.8.3
      Found dependency memutils 0.4.9
    Found dependency vibe-core 1.0.0
      Found dependency eventcore 0.8.12
        Found dependency taggedalgebraic 0.10.7
  Error getting sub package vibe-d:textfilter: Sub package "vibe-d:textfilter" doesn't exist.
  Error getting sub package vibe-d:crypto: Sub package "vibe-d:crypto" doesn't exist.
  Error getting sub package vibe-d:mongodb: Sub package "vibe-d:mongodb" doesn't exist.
  Error getting sub package vibe-d:inet: Sub package "vibe-d:inet" doesn't exist.
  Error getting sub package vibe-d:diet: Sub package "vibe-d:diet" doesn't exist.
  Error getting sub package vibe-d:web: Sub package "vibe-d:web" doesn't exist.
  Error getting sub package vibe-d:tls: Sub package "vibe-d:tls" doesn't exist.
  Found dependency vibe-d:stream 0.8.0-rc.4+commit.2.g6ae5c111
    Found dependency openssl 1.1.5+1.0.1g
    Found dependency botan 1.12.9
      Found dependency botan-math 1.0.3
  Error getting sub package vibe-d:http: Sub package "vibe-d:http" doesn't exist.
  Error getting sub package vibe-d:mail: Sub package "vibe-d:mail" doesn't exist.
Checking for upgrades.
Using cached upgrade results...
Non-optional dependency vibe-d:crypto of vibe-d not found in dependency tree!?.
```

File: stream/dub.sdl (hunk 1 above)
Comment (wilzbach) [inline stream/dub.sdl]: > Yeah, the next step would definitely be to separate this into a vibe-d:tls package. Maybe I should even better do that now... 

Don't you have to move the SSL config over to the `tls` package then?

**Comment by s-ludwig** [comment]
Err, yeah... forgot to check in tls/dub.sdl and didn't remove the old ssl dependencies. Good point ;-)

Auto-cancel is enabled, but AppVeyor is simply idle.

**Comment by s-ludwig** [comment]
Okay, I'll just rebase and merge in the hope that AppVeyor will then run on the 0.8.0 branch. If all breaks down, this can still be reverted.

**Pull request merged**